README.md
# select_modal_flutter

Flutter package to create a select with its options in a modal window with search filter.

## example

```dart
import 'package:flutter/material.dart';
import 'package:select_modal_flutter/select_modal_flutter.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Example Select Modal Flutter',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  TextEditingController controllerCity = TextEditingController();
  TextEditingController controllerCountry = TextEditingController();
  TextEditingController controllerName = TextEditingController();

  List<ItemSelect> listItemSelectCities = [
    const ItemSelect(value: 1, label: 'Caracas'),
    const ItemSelect(value: 2, label: 'San Cristobal'),
    const ItemSelect(value: 3, label: 'Merida'),
    const ItemSelect(value: 4, label: 'Maracaibo'),
    const ItemSelect(value: 4, label: 'Barquisimeto'),
    const ItemSelect(value: 4, label: 'Los Teques'),
    const ItemSelect(value: 4, label: 'La Guaira'),
  ];

  List<ItemSelect> listItemSelectNames = [
    const ItemSelect(value: 1, label: 'Pedro'),
    const ItemSelect(value: 2, label: 'Juan'),
    const ItemSelect(value: 3, label: 'Maria'),
    const ItemSelect(value: 4, label: 'Antonio'),
  ];

  bool errorSelectName = false;

  @override
  void dispose() {
    controllerCity.dispose();
    controllerCountry.dispose();
    controllerName.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Example Select Modal Flutter'),
      ),
      body: Padding(
        padding: const EdgeInsets.all(20.0),
        child: Center(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              const Text(
                '--- Select ---',
                style: TextStyle(
                  fontSize: 20.0,
                  fontWeight: FontWeight.bold,
                ),
              ),
              const SizedBox(height: 15.0),
              const Text('Names'),
              const SizedBox(
                height: 15.0,
              ),
              SelectModalFlutter(
                title: 'Name',
                searchText: 'Select names',
                controller: controllerName,
                listItemSelect: listItemSelectNames,
                borderTextField: InputBorder.none,
                boxDecoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(15.0),
                  color: Colors.white,
                  border: Border.all(color: Colors.grey, width: 1.0),
                ),
                icon: Icons.person,
                colorIcon: Colors.blue,
                error: errorSelectName,
              ),
              const SizedBox(
                height: 15.0,
              ),
              const Text(
                '--- Multi Select ---',
                style: TextStyle(
                  fontSize: 20.0,
                  fontWeight: FontWeight.bold,
                ),
              ),
              const SizedBox(height: 15.0),
              const Text('Cities'),
              const SizedBox(
                height: 15.0,
              ),
              MultiSelectModalFlutter(
                  title: 'City',
                  searchText: 'Select city',
                  listItemSelect: listItemSelectCities,
                  borderTextField: InputBorder.none,
                  boxDecoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(15.0),
                    color: Colors.white,
                    border: Border.all(color: Colors.grey, width: 1.0),
                  ),
                  colorButtonSelect: Colors.blue,
                  onItemSelect: (List<ItemSelect> items) {
                    print(items);
                  }),
              const SizedBox(height: 15.0),
              ElevatedButton(
                onPressed: () {
                  if (controllerName.text.isEmpty) {
                    setState(() {
                      errorSelectName = true;
                    });
                  } else {
                    setState(() {
                      errorSelectName = false;
                    });
                  }
                },
                child: const Text('Save'),
              )
            ],
          ),
        ),
      ),
    );
  }
}

```


CHANGELOG.md
## 0.0.1
- initial release.

## 0.0.2
- add error validation.

## 0.0.3
- add comments

## 0.1.1
- add multiselect

## 0.1.2
- fixed multiselect

## 0.1.3
- fixed select

test/select_modal_flutter_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:select_modal_flutter/select_modal_flutter.dart';

void main() {
  test('adds one to input values', () {

  });
}


lib/src/data/select.dart
import 'package:flutter/material.dart';
import 'package:select_modal_flutter/src/models/item_select.dart';
import 'package:select_modal_flutter/src/widgets/select_dialog.dart';

export 'package:select_modal_flutter/src/models/item_select.dart';

///Main class,
///[title] hintText in searh
///[searchText] optional, text in modal search,
///[onItemSelect] Function return item select
///[listItemSelect] List to search
///[boxDecoration] optional, boxdecoration to container
///[borderTextField] optiona, border to container
///[icon] optional, icon to trailing in container
///[colorIcon] optional, color icon to trailing in container
///[error] optional, error to border and icon in to container
class SelectModalFlutter extends StatefulWidget {
  final String title;
  final String? searchText;
  final Function onItemSelect;
  final List<ItemSelect> listItemSelect;
  final BoxDecoration? boxDecoration;
  final InputBorder? borderTextField;
  final IconData? icon;
  final Color? colorIcon;
  final bool? error;

  const SelectModalFlutter({
    Key? key,
    required this.title,
    required this.onItemSelect,
    this.searchText,
    required this.listItemSelect,
    this.boxDecoration,
    this.borderTextField,
    this.icon,
    this.colorIcon,
    this.error,
  }) : super(key: key);

  @override
  State<SelectModalFlutter> createState() => _SelectModalFlutterState();
}

class _SelectModalFlutterState extends State<SelectModalFlutter> {
  late ItemSelect _selectedItem = new ItemSelect();
  final controller = TextEditingController();

  @override
  void initState() {
    super.initState();
  }

  Future<void> _changeSelectOption() async {
    List<ItemSelect> filteredItems = widget.listItemSelect;
    await showDialog(
      context: context,
      useRootNavigator: false,
      builder: (context) => StatefulBuilder(
        builder: (ctx, setState) => SelectDialogWidget(
          filteredItems: filteredItems,
          searchText: widget.searchText ?? 'Buscar',
          itemList: filteredItems,
          selectedItem: _selectedItem,
          onItemChanged: (ItemSelect item) {
            _selectedItem = item;
            widget.onItemSelect(item);
            controller.text = item.label!;
            setState(() {});
          },
        ),
      ),
    );
    if (this.mounted) setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    BoxDecoration? newDecoration;
    if (widget.error != null) {
      newDecoration = widget.boxDecoration?.copyWith(
          border: widget.error == true
              ? widget.borderTextField == null
                  ? Border(bottom: BorderSide(color: Colors.red, width: 1.0))
                  : Border.all(color: Colors.red, width: 1.0)
              : null);
    }

    return Container(
      decoration: widget.boxDecoration != null
          ? widget.error != null
              ? newDecoration
              : widget.boxDecoration
          : null,
      child: Padding(
        padding: EdgeInsets.only(left: 10.0),
        child: TextField(
          onTap: _changeSelectOption,
          controller: controller,
          readOnly: true,
          decoration: InputDecoration(
            hintText: widget.title,
            border: widget.borderTextField ?? null,
            suffixIcon: widget.icon != null
                ? Icon(widget.icon,
                    color: widget.error != null
                        ? widget.error == true
                            ? Colors.red
                            : widget.colorIcon
                        : widget.colorIcon)
                : null,
          ),
        ),
      ),
    );
  }
}


lib/src/data/multi_select.dart
import 'package:flutter/material.dart';
import 'package:select_modal_flutter/src/models/item_select.dart';
import 'package:select_modal_flutter/src/widgets/widgets.dart';

export 'package:select_modal_flutter/src/models/item_select.dart';

///Main class,
///[title] hintText in searh
///[searchText] optional, text in modal search,
///[listItemSelect] List to search
///[boxDecoration] optional, boxdecoration to container
///[borderTextField] optiona, border to container
///[icon] optional, icon to trailing in container
///[colorIcon] optional, color icon to trailing in container
///[error] optional, error to border and icon in to container
///[onItemSelect] Function return list items select
///[colorButtonSelect] optional, color button select
class MultiSelectModalFlutter extends StatefulWidget {
  final String title;
  final String? searchText;
  final List<ItemSelect> listItemSelect;
  final BoxDecoration? boxDecoration;
  final InputBorder? borderTextField;
  final IconData? icon;
  final Color? colorIcon;
  final bool? error;
  final Function onItemSelect;
  final Color? colorButtonSelect;

  const MultiSelectModalFlutter({
    Key? key,
    required this.title,
    this.searchText,
    required this.listItemSelect,
    this.boxDecoration,
    this.borderTextField,
    this.icon,
    this.colorIcon,
    this.error,
    required this.onItemSelect,
    this.colorButtonSelect,
  }) : super(key: key);

  @override
  State<MultiSelectModalFlutter> createState() =>
      _MultiSelectModalFlutterState();
}

class _MultiSelectModalFlutterState extends State<MultiSelectModalFlutter> {
  late List<ItemSelect> _selectedItem = [];

  @override
  void initState() {
    super.initState();
  }

  Future<void> _changeSelectOption() async {
    List<ItemSelect> filteredItems = widget.listItemSelect;
    await showDialog(
      context: context,
      useRootNavigator: false,
      builder: (context) => StatefulBuilder(
        builder: (ctx, setState) => MultiSelectDialogWidget(
          filteredItems: filteredItems,
          searchText: widget.searchText ?? 'Buscar',
          itemList: filteredItems,
          selectedItem: _selectedItem,
          colorButtonSelect: widget.colorButtonSelect,
        ),
      ),
    ).then((value) {
      if (value is List<ItemSelect>) {
        _selectedItem = value;
        widget.onItemSelect(value);
        setState(() {});
      }
    });
    if (this.mounted) setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    BoxDecoration? newDecoration;
    if (widget.error != null) {
      newDecoration = widget.boxDecoration?.copyWith(
          border: widget.error == true
              ? widget.borderTextField == null
                  ? Border(bottom: BorderSide(color: Colors.red, width: 1.0))
                  : Border.all(color: Colors.red, width: 1.0)
              : null);
    }

    return Column(
      children: [
        Container(
          decoration: widget.boxDecoration != null
              ? widget.error != null
                  ? newDecoration
                  : widget.boxDecoration
              : null,
          child: Padding(
            padding: EdgeInsets.only(left: 10.0),
            child: TextField(
              onTap: _changeSelectOption,
              readOnly: true,
              decoration: InputDecoration(
                hintText: widget.title,
                border: widget.borderTextField ?? null,
                suffixIcon: widget.icon != null
                    ? Icon(widget.icon,
                        color: widget.error != null
                            ? widget.error == true
                                ? Colors.red
                                : widget.colorIcon
                            : widget.colorIcon)
                    : null,
              ),
            ),
          ),
        ),
        SizedBox(height: 10.0),
        SizedBox(
          width: MediaQuery.of(context).size.width - 50.0,
          child: SingleChildScrollView(
            scrollDirection: Axis.horizontal,
            child: Row(
              children: _selectedItem.map((item) => _item(item)).toList(),
            ),
          ),
        ),
      ],
    );
  }

  Widget _item(ItemSelect item) {
    return Container(
      margin: EdgeInsets.only(right: 10.0),
      padding: EdgeInsets.all(5.0),
      decoration: BoxDecoration(
        color: widget.colorButtonSelect != null
            ? widget.colorButtonSelect!.withOpacity(.2)
            : Colors.green.withOpacity(.2),
        border: Border.all(
          color: widget.colorButtonSelect != null
              ? widget.colorButtonSelect!
              : Colors.green,
        ),
        borderRadius: BorderRadius.circular(10.0),
      ),
      child: Text(item.label!),
    );
  }
}


lib/src/widgets/select_dialog.dart
import 'package:flutter/material.dart';
import 'package:select_modal_flutter/src/models/item_select.dart';

///Main widget,
///[itemList] List original to search
///[selectedItem] Item selected
///[onItemChanged] Function to return value selected
///[searchText] Text in search textfield
///[filteredItems] List filtered to search
class SelectDialogWidget extends StatefulWidget {
  final List<ItemSelect> itemList;
  final ItemSelect selectedItem;
  final ValueChanged<ItemSelect> onItemChanged;
  final String searchText;
  final List<ItemSelect> filteredItems;

  SelectDialogWidget({
    Key? key,
    required this.searchText,
    required this.itemList,
    required this.onItemChanged,
    required this.selectedItem,
    required this.filteredItems,
  }) : super(key: key);

  @override
  _SelectDialogWidgetState createState() => _SelectDialogWidgetState();
}

class _SelectDialogWidgetState extends State<SelectDialogWidget> {
  late List<ItemSelect> _filteredItems;
  late ItemSelect _selectedItem;

  @override
  void initState() {
    _selectedItem = widget.selectedItem;
    _filteredItems = widget.filteredItems;

    super.initState();
  }

  @override
  Widget build(BuildContext context) => Dialog(
        backgroundColor: Colors.white,
        child: Container(
          padding: EdgeInsets.all(10),
          child: SingleChildScrollView(
            child: Column(
              mainAxisSize: MainAxisSize.min,
              children: <Widget>[
                Padding(
                  padding: EdgeInsets.all(0),
                  child: TextField(
                    decoration: InputDecoration(
                      suffixIcon: Icon(Icons.search),
                      labelText: widget.searchText,
                    ),
                    onChanged: (value) {
                      _filteredItems = widget.itemList
                          .where((item) => item.label!
                              .toLowerCase()
                              .contains(value.toLowerCase()))
                          .toList();
                      if (this.mounted) setState(() {});
                    },
                  ),
                ),
                SizedBox(height: 20),
                Column(
                  children: _filteredItems.map((item) => _item(item)).toList(),
                ),
              ],
            ),
          ),
        ),
      );

  Widget _item(ItemSelect item) {
    return Column(
      children: [
        Container(
          color: _selectedItem.value == item.value ? Colors.grey[200] : null,
          child: ListTile(
            title: Text(
              item.label!,
              style: const TextStyle(fontWeight: FontWeight.w700),
            ),
            trailing:
                _selectedItem.value == item.value ? Icon(Icons.check) : null,
            onTap: () {
              widget.onItemChanged(item);
              Navigator.of(context).pop();
            },
          ),
        ),
        const Divider(thickness: 1)
      ],
    );
  }
}


lib/src/widgets/multi_select_dialog.dart
import 'package:flutter/material.dart';
import 'package:select_modal_flutter/src/models/item_select.dart';

///Main widget,
///[itemList] List original to search
///[selectedItem] List items selected
///[onItemChanged] Function to return value selected
///[searchText] Text in search textfield
///[filteredItems] List filtered to search
class MultiSelectDialogWidget extends StatefulWidget {
  final List<ItemSelect> itemList;
  final List<ItemSelect> selectedItem;
  final String searchText;
  final List<ItemSelect> filteredItems;
  final Color? colorButtonSelect;

  MultiSelectDialogWidget({
    Key? key,
    required this.searchText,
    required this.itemList,
    required this.selectedItem,
    required this.filteredItems,
    this.colorButtonSelect,
  }) : super(key: key);

  @override
  _MultiSelectDialogWidgetState createState() =>
      _MultiSelectDialogWidgetState();
}

class _MultiSelectDialogWidgetState extends State<MultiSelectDialogWidget> {
  late List<ItemSelect> _filteredItems;
  late List<ItemSelect> _selectedItem;

  @override
  void initState() {
    _selectedItem = [...widget.selectedItem];
    _filteredItems = widget.filteredItems;

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    final width = MediaQuery.of(context).size.width;
    return Dialog(
      backgroundColor: Colors.white,
      child: Container(
        padding: EdgeInsets.all(10),
        child: SingleChildScrollView(
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              Padding(
                padding: EdgeInsets.all(0),
                child: Row(
                  children: [
                    SizedBox(
                      width: _selectedItem.isNotEmpty
                          ? width - 140.0
                          : width - 100.0,
                      child: TextField(
                        decoration: InputDecoration(
                          suffixIcon: Icon(Icons.search),
                          labelText: widget.searchText,
                        ),
                        onChanged: (value) {
                          _filteredItems = widget.itemList
                              .where((item) => item.label!
                                  .toLowerCase()
                                  .contains(value.toLowerCase()))
                              .toList();
                          if (this.mounted) setState(() {});
                        },
                      ),
                    ),
                    if (_selectedItem.isNotEmpty) _iconSave()
                  ],
                ),
              ),
              SizedBox(height: 20),
              Column(
                children: _filteredItems.map((item) => _item(item)).toList(),
              ),
            ],
          ),
        ),
      ),
    );
  }

  Widget _iconSave() {
    return GestureDetector(
      onTap: () {
        Navigator.pop(context, _selectedItem);
      },
      child: Container(
        padding: EdgeInsets.all(5.0),
        decoration: BoxDecoration(
          color: widget.colorButtonSelect != null
              ? widget.colorButtonSelect!.withOpacity(.2)
              : Colors.green.withOpacity(.2),
          border: Border.all(
            color: widget.colorButtonSelect != null
                ? widget.colorButtonSelect!
                : Colors.green,
          ),
          borderRadius: BorderRadius.circular(10.0),
        ),
        child: Icon(Icons.check),
      ),
    );
  }

  Widget _item(ItemSelect item) {
    return Column(
      children: [
        Container(
          color: _selectedItem.contains(item) ? Colors.grey[200] : null,
          child: ListTile(
            title: Text(
              item.label!,
              style: const TextStyle(fontWeight: FontWeight.w700),
            ),
            trailing: _selectedItem.contains(item) ? Icon(Icons.check) : null,
            onTap: () {
              final elementExists = _selectedItem.contains(item);
              if (!elementExists) {
                _selectedItem.add(item);
              } else {
                var indice = _selectedItem.indexOf(item);
                this._selectedItem.removeAt(indice);
              }
              setState(() {});
            },
          ),
        ),
        const Divider(thickness: 1)
      ],
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:select_modal_flutter/select_modal_flutter.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Example Select Modal Flutter',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  TextEditingController controllerCity = TextEditingController();
  TextEditingController controllerCountry = TextEditingController();
  TextEditingController controllerName = TextEditingController();

  List<ItemSelect> listItemSelectCities = [
    const ItemSelect(value: 1, label: 'Caracas'),
    const ItemSelect(value: 2, label: 'San Cristobal'),
    const ItemSelect(value: 3, label: 'Merida'),
    const ItemSelect(value: 4, label: 'Maracaibo'),
    const ItemSelect(value: 4, label: 'Barquisimeto'),
    const ItemSelect(value: 4, label: 'Los Teques'),
    const ItemSelect(value: 4, label: 'La Guaira'),
  ];

  List<ItemSelect> listItemSelectNames = [
    const ItemSelect(value: 1, label: 'Pedro'),
    const ItemSelect(value: 2, label: 'Juan'),
    const ItemSelect(value: 3, label: 'Maria'),
    const ItemSelect(value: 4, label: 'Antonio'),
  ];

  bool errorSelectName = false;

  @override
  void dispose() {
    controllerCity.dispose();
    controllerCountry.dispose();
    controllerName.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Example Select Modal Flutter'),
      ),
      body: Padding(
        padding: const EdgeInsets.all(20.0),
        child: Center(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              const Text(
                '--- Select ---',
                style: TextStyle(
                  fontSize: 20.0,
                  fontWeight: FontWeight.bold,
                ),
              ),
              const SizedBox(height: 15.0),
              const Text('Names'),
              const SizedBox(
                height: 15.0,
              ),
              SelectModalFlutter(
                title: 'Name',
                searchText: 'Select names',
                onItemSelect: (ItemSelect value) {
                  print(value);
                },
                listItemSelect: listItemSelectNames,
                borderTextField: InputBorder.none,
                boxDecoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(15.0),
                  color: Colors.white,
                  border: Border.all(color: Colors.grey, width: 1.0),
                ),
                icon: Icons.person,
                colorIcon: Colors.blue,
                error: errorSelectName,
              ),
              const SizedBox(
                height: 15.0,
              ),
              const Text(
                '--- Multi Select ---',
                style: TextStyle(
                  fontSize: 20.0,
                  fontWeight: FontWeight.bold,
                ),
              ),
              const SizedBox(height: 15.0),
              const Text('Cities'),
              const SizedBox(
                height: 15.0,
              ),
              MultiSelectModalFlutter(
                  title: 'City',
                  searchText: 'Select city',
                  listItemSelect: listItemSelectCities,
                  borderTextField: InputBorder.none,
                  boxDecoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(15.0),
                    color: Colors.white,
                    border: Border.all(color: Colors.grey, width: 1.0),
                  ),
                  colorButtonSelect: Colors.blue,
                  onItemSelect: (List<ItemSelect> items) {
                    print(items);
                  }),
              const SizedBox(height: 15.0),
              ElevatedButton(
                onPressed: () {
                  if (controllerName.text.isEmpty) {
                    setState(() {
                      errorSelectName = true;
                    });
                  } else {
                    setState(() {
                      errorSelectName = false;
                    });
                  }
                },
                child: const Text('Save'),
              )
            ],
          ),
        ),
      ),
    );
  }
}
